### ISSUE TITLE
Lint `PhantomData::default()` -> `PhantomData`

### ISSUE THREAD

**Issue by alecmocatta** [rust-lang/rust-clippy#10609]
### What it does

`PhantomData::default()` is quite common https://github.com/search?q=language%3ARust+%22PhantomData%3A%3Adefault%28%29%22&type=code&ref=advsearch though of course the `::default()` is actually redundant.

Perhaps a lint that suggests removing the `::default()` would be useful?.

### Category

style, pedantic

### Advantage

Shorter, ostensibly avoids a function call.

### Drawbacks

Perhaps

```rust
Foo {
    _marker: PhantomData::default()
}
```

makes more sense to some people than the slightly special

```rust
Foo {
    _marker: PhantomData
}
```

### Example

```rust
Foo {
    _marker: PhantomData::default()
}
```

Could be written as:

```rust
Foo {
    _marker: PhantomData
}
```

**Comment by llogiq** [comment]
I think it should be `complexity` and should not lint in macros or if the PhantomData type is introduced by generics.

**Comment by Alexendoo** [comment]
Could be extended to other unit structs too

**Comment by Alexendoo** [comment]
To be conservative the lint can check the `Default` impl with https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/context/struct.TyCtxt.html#method.is_builtin_derived

**Comment by llogiq** [comment]
@asquared31415 how many different unit struct instances do we expect per type (absent generics perhaps)?

**Comment by asquared31415** [comment]
I am talking about a case where the body of `Default::default` does something with side effects that is important.

However, checking if `Default` is automatically derived is sufficient to catch most normal empty structs and that derive is always well behaved, so that check solves any potential issues that I can think of.

**Comment by schubart** [comment]
I don't think checking if `Default` is automatically derived is good enough: It might be trivial and automatically derived *today*, but the owner of the struct might change it to a hand-crafted implementation in the future.

**Comment by schubart** [comment]
Apologies, I don't have time to work on this at the moment. I don't want to "hog" this good-first-issue, so I'll unassign myself and give someone else the chance to claim it.

**Comment by Alexendoo** [comment]
A `Default` impl of a unit struct having side effects is already bizarre/unlikely enough that I would be surprised to find any real world examples, I wouldn't worry about a derive changing to one in a future version

**Comment by llogiq** [comment]
Agreed. I would even go so far as to say the additional check isn't needed. By convention, `Default::default()` should return a stand-in value, and if the type is public, it cannot rely on not having `Default` called to construct an instance; if the code wants to do so, it can add `#[non_exhaustive]` or a private unit field. So for public unit structs, `T` and `T::default()` are for all intents and purposes the same.
